roachtest: tpchbench/tpch/nodes=3/cpu=4/sf=1 failed

[(roachtest).tpchbench/tpch/nodes=3/cpu=4/sf=1 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1826340&tab=buildLog) on [release-19.2@930c0fad9f0c51f258417b4773138e7b8066c3c1](https://github.com/cockroachdb/cockroach/commits/930c0fad9f0c51f258417b4773138e7b8066c3c1):

```
		    github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s returned:
		    stderr:
		    %s
		    stdout:
		    %s
		    -- arg 1: <string>
		    -- arg 2: <string>
		    -- arg 3: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1826340-1585118678-01-n4cpu4:4 -- ./workload run querybench --db=tpch --concurrency=1 --query-file=tpch --num-runs=3 --max-ops=72 {pgurl:1-3} --histograms=perf/stats.json --histograms-max-latency=8m20s returned:
		    stderr:
		    Error: failed to prepare query "SELECT s_suppkey, s_name, s_address, s_phone, total_revenue FROM supplier, revenue0 WHERE s_suppkey = supplier_no AND total_revenue = ( SELECT max(total_revenue) FROM revenue0) ORDER BY s_suppkey": pq: relation "revenue0" does not exist
		    Error:  exit status 1
		    
		    stdout::
		  - exit status 1

	cluster.go:2368,tpchbench.go:120,tpchbench.go:245,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.runTPCHBench
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchbench.go:120
		    main.registerTPCHBenchSpec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchbench.go:245
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - error with attached stack trace:
		    main.(*monitor).wait.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2412
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor task failed:
		  - error with attached stack trace:
		    main.init
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2309
		    runtime.doInit
		    	/usr/local/go/src/runtime/proc.go:5222
		    runtime.main
		    	/usr/local/go/src/runtime/proc.go:190
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - Goexit() was called

```

<details><summary>More</summary><p>


Artifacts: [/tpchbench/tpch/nodes=3/cpu=4/sf=1](https://teamcity.cockroachdb.com/viewLog.html?buildId=1826340&tab=artifacts#/tpchbench/tpch/nodes=3/cpu=4/sf=1)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchbench%2Ftpch%2Fnodes%3D3%2Fcpu%3D4%2Fsf%3D1.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchbench/tpch/nodes=3/cpu=4/sf=1 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1829298&tab=buildLog) on [release-19.2@d105bee7df2a29942612e946a3606f146baa6b8a](https://github.com/cockroachdb/cockroach/commits/d105bee7df2a29942612e946a3606f146baa6b8a):

```
		    stderr:
		    %s
		    stdout:
		    %s
		    -- arg 1: <string>
		    -- arg 2: <string>
		    -- arg 3: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1829298-1585205017-06-n4cpu4:4 -- ./workload run querybench --db=tpch --concurrency=1 --query-file=tpch --num-runs=3 --max-ops=72 {pgurl:1-3} --histograms=perf/stats.json --histograms-max-latency=8m20s returned:
		    stderr:
		    Error: failed to prepare query "SELECT s_suppkey, s_name, s_address, s_phone, total_revenue FROM supplier, revenue0 WHERE s_suppkey = supplier_no AND total_revenue = ( SELECT max(total_revenue) FROM revenue0) ORDER BY s_suppkey": pq: relation "revenue0" does not exist
		    Error: DEAD_ROACH_PROBLEM:
		      - error with user detail: Node 4. Command with error:
		        ```
		        ./workload run querybench --db=tpch --concurrency=1 --query-file=tpch --num-runs=3 --max-ops=72 {pgurl:1-3} --histograms=perf/stats.json --histograms-max-latency=8m20s
		        ```
		      - exit status 1
		    
		    stdout::
		  - exit status 30

	cluster.go:2368,tpchbench.go:120,tpchbench.go:245,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.runTPCHBench
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchbench.go:120
		    main.registerTPCHBenchSpec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchbench.go:245
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - error with attached stack trace:
		    main.(*monitor).wait.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2412
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor task failed:
		  - error with attached stack trace:
		    main.init
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2309
		    runtime.doInit
		    	/usr/local/go/src/runtime/proc.go:5222
		    runtime.main
		    	/usr/local/go/src/runtime/proc.go:190
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - Goexit() was called

```

<details><summary>More</summary><p>


Artifacts: [/tpchbench/tpch/nodes=3/cpu=4/sf=1](https://teamcity.cockroachdb.com/viewLog.html?buildId=1829298&tab=artifacts#/tpchbench/tpch/nodes=3/cpu=4/sf=1)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchbench%2Ftpch%2Fnodes%3D3%2Fcpu%3D4%2Fsf%3D1.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchbench/tpch/nodes=3/cpu=4/sf=1 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1832643&tab=buildLog) on [release-19.2@be4d412e235f84b5fae712c0d69fa285ec16641e](https://github.com/cockroachdb/cockroach/commits/be4d412e235f84b5fae712c0d69fa285ec16641e):

```
		    stderr:
		    %s
		    stdout:
		    %s
		    -- arg 1: <string>
		    -- arg 2: <string>
		    -- arg 3: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1832643-1585291392-11-n4cpu4:4 -- ./workload run querybench --db=tpch --concurrency=1 --query-file=tpch --num-runs=3 --max-ops=72 {pgurl:1-3} --histograms=perf/stats.json --histograms-max-latency=8m20s returned:
		    stderr:
		    Error: failed to prepare query "SELECT s_suppkey, s_name, s_address, s_phone, total_revenue FROM supplier, revenue0 WHERE s_suppkey = supplier_no AND total_revenue = ( SELECT max(total_revenue) FROM revenue0) ORDER BY s_suppkey": pq: relation "revenue0" does not exist
		    Error: DEAD_ROACH_PROBLEM:
		      - error with user detail: Node 4. Command with error:
		        ```
		        ./workload run querybench --db=tpch --concurrency=1 --query-file=tpch --num-runs=3 --max-ops=72 {pgurl:1-3} --histograms=perf/stats.json --histograms-max-latency=8m20s
		        ```
		      - exit status 1
		    
		    stdout::
		  - exit status 30

	cluster.go:2368,tpchbench.go:120,tpchbench.go:245,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.runTPCHBench
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchbench.go:120
		    main.registerTPCHBenchSpec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchbench.go:245
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - error with attached stack trace:
		    main.(*monitor).wait.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2412
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor task failed:
		  - error with attached stack trace:
		    main.init
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2309
		    runtime.doInit
		    	/usr/local/go/src/runtime/proc.go:5222
		    runtime.main
		    	/usr/local/go/src/runtime/proc.go:190
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - Goexit() was called

```

<details><summary>More</summary><p>


Artifacts: [/tpchbench/tpch/nodes=3/cpu=4/sf=1](https://teamcity.cockroachdb.com/viewLog.html?buildId=1832643&tab=artifacts#/tpchbench/tpch/nodes=3/cpu=4/sf=1)
Related:
- #46607 roachtest: tpchbench/tpch/nodes=3/cpu=4/sf=1 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-master](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-master) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchbench%2Ftpch%2Fnodes%3D3%2Fcpu%3D4%2Fsf%3D1.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/workload/querybench/query_bench.go
+++ b/pkg/workload/querybench/query_bench.go
@@ -134,16 +134,17 @@ func (g *queryBench) Ops(urls []string, reg *histogram.Registry) (workload.Query
 
 	stmts := make([]namedStmt, len(g.queries))
 	for i, query := range g.queries {
-		stmt, err := db.Prepare(query)
-		if err != nil {
-			return workload.QueryLoad{}, errors.Wrapf(err, "failed to prepare query %q", query)
-		}
 		stmts[i] = namedStmt{
 			// TODO(solon): Allow specifying names in the query file rather than using
 			// the entire query as the name.
 			name: fmt.Sprintf("%2d: %s", i+1, query),
-			stmt: stmt,
 		}
+		stmt, err := db.Prepare(query)
+		if err != nil {
+			stmts[i].query = query
+			continue
+		}
+		stmts[i].preparedStmt = stmt
 	}
 
 	maxNumStmts := 0
@@ -192,7 +193,11 @@ func GetQueries(path string) ([]string, error) {
 
 type namedStmt struct {
 	name string
-	stmt *gosql.Stmt
+	// We will try to Prepare the statement, and if that succeeds, the prepared
+	// statement will be stored in `preparedStmt', otherwise, we will store
+	// plain query in 'query'.
+	preparedStmt *gosql.Stmt
+	query        string
 }
 
 type queryBenchWorker struct {
@@ -221,15 +226,31 @@ func (o *queryBenchWorker) run(ctx context.Context) error {
 	stmt := o.stmts[o.stmtIdx%len(o.stmts)]
 	o.stmtIdx++
 
-	rows, err := stmt.stmt.Query()
-	if err != nil {
-		return err
+	exhaustRows := func(execFn func() (*gosql.Rows, error)) error {
+		rows, err := execFn()
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+		for rows.Next() {
+		}
+		if err := rows.Err(); err != nil {
+			return err
+		}
+		return nil
 	}
-	defer rows.Close()
-	for rows.Next() {
-	}
-	if err := rows.Err(); err != nil {
-		return err
+	if stmt.preparedStmt != nil {
+		if err := exhaustRows(func() (*gosql.Rows, error) {
+			return stmt.preparedStmt.Query()
+		}); err != nil {
+			return err
+		}
+	} else {
+		if err := exhaustRows(func() (*gosql.Rows, error) {
+			return o.db.Query(stmt.query)
+		}); err != nil {
+			return err
+		}
 	}
 	elapsed := timeutil.Since(start)
 	if o.verbose {

--- a/pkg/workload/querybench/tpch-queries
+++ b/pkg/workload/querybench/tpch-queries
@@ -44,9 +44,7 @@ SELECT c_count, count(*) AS custdist FROM ( SELECT c_custkey, count(o_orderkey) 
 SELECT 100.00 * sum(CASE WHEN p_type LIKE 'PROMO%' THEN l_extendedprice * (1 - l_discount) ELSE 0 END) / sum(l_extendedprice * (1 - l_discount)) AS promo_revenue FROM lineitem, part WHERE l_partkey = p_partkey AND l_shipdate >= DATE '1995-09-01' AND l_shipdate < DATE '1995-09-01' + INTERVAL '1' MONTH
 
 -- query 15
-CREATE VIEW revenue0 (supplier_no, total_revenue) AS SELECT l_suppkey, sum(l_extendedprice * (1 - l_discount)) FROM lineitem WHERE l_shipdate >= DATE '1996-01-01' AND l_shipdate < DATE '1996-01-01' + INTERVAL '3' MONTH GROUP BY l_suppkey
-SELECT s_suppkey, s_name, s_address, s_phone, total_revenue FROM supplier, revenue0 WHERE s_suppkey = supplier_no AND total_revenue = ( SELECT max(total_revenue) FROM revenue0) ORDER BY s_suppkey
-DROP VIEW revenue0
+CREATE VIEW revenue0 (supplier_no, total_revenue) AS SELECT l_suppkey, sum(l_extendedprice * (1 - l_discount)) FROM lineitem WHERE l_shipdate >= DATE '1996-01-01' AND l_shipdate < DATE '1996-01-01' + INTERVAL '3' MONTH GROUP BY l_suppkey; SELECT s_suppkey, s_name, s_address, s_phone, total_revenue FROM supplier, revenue0 WHERE s_suppkey = supplier_no AND total_revenue = ( SELECT max(total_revenue) FROM revenue0) ORDER BY s_suppkey; DROP VIEW revenue0
 
 -- query 16
 SELECT p_brand, p_type, p_size, count(distinct ps_suppkey) AS supplier_cnt FROM partsupp, part WHERE p_partkey = ps_partkey AND p_brand <> 'Brand#45' AND p_type NOT LIKE 'MEDIUM POLISHED%' AND p_size IN (49, 14, 23, 45, 19, 3, 36, 9) AND ps_suppkey NOT IN ( SELECT s_suppkey FROM supplier WHERE s_comment LIKE '%Customer%Complaints%') GROUP BY p_brand, p_type, p_size ORDER BY supplier_cnt DESC, p_brand, p_type, p_size

